### PR TITLE
fuzz: Fix difficulty target generation in `p2p_headers_presync`

### PR DESCRIPTION
**Pull request by marcofleon**
In the `p2p_headers_presync` fuzz target, this assertion failed:
```
 assert(total_work < chainman.MinimumChainWork());
 ```
Input that triggered the failure: [p2ppresync_crash.txt](https://github.com/user-attachments/files/17620203/p2ppresync_crash.txt)

The test previously used `ConsumeIntegralInRange` to generate header difficulty targets within a hardcoded range. The fuzzer found specific values in that range that correspond to very low thresholds due to how [`SetCompact`][setcompact-link] works. The total work of a long enough test chain ended up exceeding `MinimumChainWork`.

Fix this by adding a new `ConsumeArithUInt256InRange` helper function and use it in the fuzz test to generate target values within the originally intended range. The target is then converted to an `nBits` value using `GetCompact()`.

For some more context, see https://github.com/bitcoin/bitcoin/pull/30918.

[setcompact-link]: https://github.com/bitcoin/bitcoin/blob/6463117a29294f6ddc9fafecfd1e9023956cc41b/src/arith_uint256.h#L251-L271

### CODE REVIEW

**Comment by instagibbs** [comment]
To be clear, `nBits` is "deterministic" in that the value, given the chain history, is static, right?

**Comment by marcofleon** [comment]
> To be clear, `nBits` is "deterministic" in that the value, given the chain history, is static, right?

Hmm I'm not entirely sure what you're asking. `nBits` does always start with the genesis block value, yes. Then it'll either change or stay the same according to the fuzzer. Is that what you meant?

**Comment by instagibbs** [comment]
@marcofleon Ah, I didn't understand how pre-sync nBits checking works, my question makes no sense for pre-sync

File: src/test/fuzz/util.h [hunk 1]
```diff
@@ -180,6 +180,20 @@ template <typename WeakEnumType, size_t size>
...
+    assert(min <= max);
+    const arith_uint256 range = max - min;
+    const arith_uint256 value = ConsumeArithUInt256(fuzzed_data_provider);
+    arith_uint256 result = value;
+    // Avoid division by 0, in case range + 1 results in overflow.
+    if (range != ~arith_uint256(0)) {
+        const arith_uint256 quotient = value / (range + 1);
+        result = value - (quotient * (range + 1));
+    }
+    return min + result;
```
Comment (instagibbs) [inline src/test/fuzz/util.h:194]: some pedantic assertions to catch hypothetical issues with the function earlier
```suggestion
    assert(min + result >= min);
    assert(min + result <= max);
    return min + result;
```

File: src/test/fuzz/p2p_headers_presync.cpp [hunk 2]
```diff
@@ -90,9 +91,14 @@ CBlockHeader ConsumeHeader(FuzzedDataProvider& fuzzed_data_provider, const uint2
     CBlockHeader header;
     header.nNonce = 0;
     // Either use the previous difficulty or let the fuzzer choose
-    header.nBits = fuzzed_data_provider.ConsumeBool() ?
-                       prev_nbits :
-                       fuzzed_data_provider.ConsumeIntegralInRange<uint32_t>(0x17058EBE, 0x1D00FFFF);
+    if (fuzzed_data_provider.ConsumeBool()) {
+        header.nBits = prev_nbits;
+    } else {
+        arith_uint256 lower_target = UintToArith256(uint256{"000000000000000000058ebe0000000000000000000000000000000000000000"});
```
Comment (instagibbs) [inline src/test/fuzz/p2p_headers_presync.cpp:97]: are there any easy ways to verify these lower and higher target values?

Comment (marcofleon) [reply src/test/fuzz/p2p_headers_presync.cpp:97]: The upper target is the `nBits` value from the genesis block (0x1D00FFFF) and the lower one is the `nBits` from some mainnet block at a certain height (0x17058EBE). Using `SetCompact` on those values gives those two uint256 values. The work returned from the lower target multiplied by the longest possible chain (1600) in the test should be lower than `MinimumChainWork`. 

I get 0x11fe50e0ab9cbcf864c0500 for the theoretical max amount of work a headers chain from this test could have. `MinimumChainWork` is 0x88e186b70e0862c193ec44d6.

File: src/test/fuzz/util.h (hunk 1 above)
Comment (marcofleon) [reply src/test/fuzz/util.h]: Done.

File: src/test/fuzz/p2p_headers_presync.cpp (hunk 2 above)
Comment (marcofleon) [reply src/test/fuzz/p2p_headers_presync.cpp:97]: I should also add, I used the expression in `GetBlockProof` https://github.com/bitcoin/bitcoin/blob/65b194193661e27cf2d9c0e0d7e3b627a379513a/src/chain.cpp#L143 to calculate the work on a single header.

Comment (marcofleon) [reply src/test/fuzz/p2p_headers_presync.cpp:97]: If I wanted to do a lower target based on `MinimumChainWork`, I think I could do
```
target = (~work + 1) / work
```
where work would be `MinimumChainWork / 1600` or maybe a bit less to be safe.

Comment (instagibbs) [reply src/test/fuzz/p2p_headers_presync.cpp:97]: Would this be correct? I would like this to be easily verifiable, and filling in the FIXMEs would make that simple
```
diff --git a/src/test/fuzz/p2p_headers_presync.cpp b/src/test/fuzz/p2p_headers_presync.cpp
index 08e8d08c08..41ac8d172b 100644
--- a/src/test/fuzz/p2p_headers_presync.cpp
+++ b/src/test/fuzz/p2p_headers_presync.cpp
@@ -92,12 +92,14 @@ CBlockHeader ConsumeHeader(FuzzedDataProvider& fuzzed_data_provider, const uint2
     header.nNonce = 0;
     // Either use the previous difficulty or let the fuzzer choose
     if (fuzzed_data_provider.ConsumeBool()) {
         header.nBits = prev_nbits;
     } else {
-        arith_uint256 lower_target = UintToArith256(uint256{"000000000000000000058ebe0000000000000000000000000000000000000000"});
-        arith_uint256 upper_target = UintToArith256(uint256{"00000000ffff0000000000000000000000000000000000000000000000000000"});
+        arith_uint256 lower_target;
+        lower_target.SetCompact(0x17058ebe); // "bits" result for testnet3 getblockheader <FIXME hash>
+        arith_uint256 upper_target;
+        upper_target.SetCompact(0x1d00ffff); // "bits" result for testnet3 getblockheader <FIXME hash>
         arith_uint256 target = ConsumeArithUInt256InRange(fuzzed_data_provider, lower_target, upper_target);
         header.nBits = target.GetCompact();
     }
     header.nTime = ConsumeTime(fuzzed_data_provider);
     header.hashPrevBlock = prev_hash;
```

Comment (marcofleon) [reply src/test/fuzz/p2p_headers_presync.cpp]: Agreed, it should be clear where the hardcoded values are coming from. It doesn't seem necessary to call `SetCompact` every iteration, so I decided to explain in the comment above. Let me know what you think.

File: src/test/fuzz/p2p_headers_presync.cpp [hunk 3]
```diff
@@ -89,10 +90,19 @@ CBlockHeader ConsumeHeader(FuzzedDataProvider& fuzzed_data_provider, const uint2
...
     header.nNonce = 0;
-    // Either use the previous difficulty or let the fuzzer choose
-    header.nBits = fuzzed_data_provider.ConsumeBool() ?
-                       prev_nbits :
-                       fuzzed_data_provider.ConsumeIntegralInRange<uint32_t>(0x17058EBE, 0x1D00FFFF);
+    // Either use the previous difficulty or let the fuzzer choose. The upper target in the
+    // range comes from the bits value of the genesis block, which is 0x1d00ffff. The lower
+    // target comes from the bits value of mainnet block 840000, which is 0x17034219.
+    // Calling lower_target.SetCompact(0x17034219) and upper_target.SetCompact(0x1d00ffff)
+    // should return the values below.
```
[the remark this answers is not in the document]
Comment (marcofleon) [reply src/test/fuzz/p2p_headers_presync.cpp:97]: Good call, done.

File: src/test/fuzz/p2p_headers_presync.cpp [hunk 4]
```diff
@@ -89,10 +90,23 @@ CBlockHeader ConsumeHeader(FuzzedDataProvider& fuzzed_data_provider, const uint2
 {
     CBlockHeader header;
     header.nNonce = 0;
-    // Either use the previous difficulty or let the fuzzer choose
-    header.nBits = fuzzed_data_provider.ConsumeBool() ?
-                       prev_nbits :
-                       fuzzed_data_provider.ConsumeIntegralInRange<uint32_t>(0x17058EBE, 0x1D00FFFF);
+    // Either use the previous difficulty or let the fuzzer choose. The upper target in the
```
Comment (instagibbs) [inline src/test/fuzz/p2p_headers_presync.cpp:93]: maybe leave a note that in presync we're allowed to have arbitrary nBits adjustments, within the clamped values?

**Comment by instagibbs** [comment]
ACK https://github.com/bitcoin/bitcoin/pull/31213/commits/a6ca8f324396522e9748c9a7bbefb3bf1c74a436

seems strictly better than as-is

**Review by dergoegge — APPROVED** [review]
Code review ACK a6ca8f324396522e9748c9a7bbefb3bf1c74a436

File: src/test/fuzz/p2p_headers_presync.cpp (hunk 4 above)
Comment (marcofleon) [reply src/test/fuzz/p2p_headers_presync.cpp:93]: As is, the test doesn't hit that part in `pow.cpp`. But I'll look to improve this test and its coverage in a future PR. Still thinking about the best way to do it. For now, I'm going leave it as is and just be sure that the crash is fixed.

**Review by brunoerg — APPROVED** [review]
code review ACK a6ca8f324396522e9748c9a7bbefb3bf1c74a436

**Pull request merged**